### PR TITLE
Activity tracker

### PR DESCRIPTION
**Pull request by priyanka-TL**

### CODE REVIEW

File: src/controllers/v1/actions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	 * @name update
+	 * @param {Integer} id  action id.
+	 * @param {Object} body  action data
+	 * @returns {JSON} - Detail of action as response.
+	 */
+	async update(req) {
+		try {
+			let action
+			if (req.params.id) {
+				action = await actionService.update(req.params.id ? req.params.id : '', req.body)
```
Comment (adithyadinesh0412) [inline src/controllers/v1/actions.js]: We have `if (req.params.id)` check. so again a ternary operator in the function call is redundant. @priyanka-TL

File: src/database/migrations/20240815181652-add-permission-for-actions.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+					module: 'actions',
+					request_type: ['POST', 'GET'],
+					api_path: '/scp/v1/actions/*',
+					created_at: new Date(),
+					updated_at: new Date(),
+					created_by: 0,
+				},
+				{
+					role_title: common.ORG_ADMIN_ROLE,
+					permission_id: await getPermissionId('actions', ['POST', 'GET'], '/scp/v1/actions/*'),
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240815181652-add-permission-for-actions.js]: Should the ORG_ADMIN have access to actions apis? 
Since it has no org based flows and common for all organization should we give this access to the org_admins?
@priyanka-TL

Comment (priyanka-TL) [reply src/database/migrations/20240815181652-add-permission-for-actions.js]: its same like modules org admin also can add their own actions

File: src/validators/v1/actions.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		req.checkParams('id')
+			.trim()
+			.optional()
+			.notEmpty()
+			.withMessage('id param is empty')
+			.isNumeric()
+			.withMessage('id param is invalid, must be an integer')
+
+		req.checkBody('code')
+			.notEmpty()
```
Comment (adithyadinesh0412) [inline src/validators/v1/actions.js:26]: Since we are passing id , can we make code optional?

Comment (priyanka-TL) [reply src/validators/v1/actions.js:26]: Whatever database field is mandatory to create record and it doesn't have default values, we can add that as a mandatory param in req body. So it wont throw error while creation or updation

File: src/database/migrations/20240813123008-create-actions-table.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+'use strict'
+
+/** @type {import('sequelize-cli').Migration} */
+module.exports = {
+	async up(queryInterface, Sequelize) {
+		await queryInterface.createTable('actions', {
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240813123008-create-actions-table.js:6]: Can we add paranoid to do a soft delete in this table? @priyanka-TL

File: src/database/migrations/20240814125518-add-default-actions.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+				updated_at: timestamp,
+			}
+		})
+
+		// Insert the data into the 'actions' table
+		await queryInterface.bulkInsert('actions', actionsFinalArray)
+	},
+
+	down: async (queryInterface, Sequelize) => {
+		await queryInterface.bulkDelete('actions', null, {})
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240814125518-add-default-actions.js:65]: Can we delete only specific items we added while reverting the migration?
In case we are reverting the data added after a point where we have actual user created data , we might loose it. 
@priyanka-TL

File: src/database/migrations/20240815181652-add-permission-for-actions.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+					permission_id: await getPermissionId('actions', ['POST', 'GET', 'DELETE'], '/scp/v1/actions/*'),
+					module: 'actions',
+					request_type: ['POST', 'GET', 'DELETE'],
+					api_path: '/scp/v1/actions/*',
+					created_at: new Date(),
+					updated_at: new Date(),
+					created_by: 0,
+				},
+				{
+					role_title: common.ORG_ADMIN_ROLE,
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240815181652-add-permission-for-actions.js:81]: Should we give delete / update permission to ORG_ADMIN? Since the activities are not org centric and common for the entire system. @priyanka-TL

File: src/database/models/activities.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+			},
+		},
+		{
+			modelName: 'Activity',
+			tableName: 'activities',
+			freezeTableName: true,
+			paranoid: true,
+			indexes: [
+				{
+					fields: ['organization_id', 'user_id'],
```
Comment (adithyadinesh0412) [inline src/database/models/activities.js]: Should we index action id also? In case if we have to search based on any specific action like created / deleted ? @priyanka-TL

File: src/helpers/requester.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+const post = async (baseURL, route = '', headers = {}, body = {}, queryParams = {}) => {
+	try {
+		baseURL = baseURL.replace(/\/$/, '')
+		let url = baseURL + route
+		if (!utils.isEmpty(queryParams)) url += '?' + new URLSearchParams(queryParams).toString()
+		console.log('URL: ', url)
+		console.log('HEADERS: ', JSON.stringify(headers, null, 2))
+		console.log('BODY: ', JSON.stringify(body, null, 2))
+		const response = await axios.post(url, body, { headers, timeout: 3000 })
```
Comment (adithyadinesh0412) [inline src/helpers/requester.js]: can we have the timeout value from common? @priyanka-TL

File: src/database/migrations/20240814125518-add-default-actions.js (hunk 5 above)
Comment (priyanka-TL) [reply src/database/migrations/20240814125518-add-default-actions.js:65]: @adithyadinesh0412 As it a first migration, we are adding the first data hence we can bulk delete all. If we are adding one action then in the down action we can delete only that time,

File: src/database/migrations/20240815181652-add-permission-for-actions.js (hunk 6 above)
Comment (priyanka-TL) [reply src/database/migrations/20240815181652-add-permission-for-actions.js:81]: @adithyadinesh0412  That is not required, its not org centric, but still org admin can create new and enhance or delete one item and instead of that they can add new item. Its there choice, only thing is if they are adding new actions they need to write the code for that

File: src/helpers/requester.js
Comment (adithyadinesh0412) [inline src/helpers/requester.js:1]: We have a file generic/request.js which handles post and get requests. How is this different from that? @priyanka-TL

File: src/database/models/activities.js (hunk 7 above)
Comment (priyanka-TL) [reply src/database/models/activities.js]: We cannot delete the activity of any user. Only thing is we can list it, For list as of now org and user id is enough

Comment (priyanka-TL) [reply src/database/models/activities.js]: Its like a tracker, we cant delete anything

File: src/helpers/requester.js
Comment (priyanka-TL) [reply src/helpers/requester.js:1]: It is specific to events

File: src/database/migrations/20240815181652-add-permission-for-actions.js (hunk 6 above)
Comment (adithyadinesh0412) [reply src/database/migrations/20240815181652-add-permission-for-actions.js:81]: @aks30  Having a confusion would like to have your thoughts here.
So here we are giving the permission to ORG_ADMINs to do CRUD operations on Actions table. 
Since actions are not org centric I'm doubting if we have to do that or not. 
My point is; what if an org admin of `A` changes the action description from one language to other or any text change specific to his/her org. It will be reflected throughout the actions in the system which is not right in a SAAS model. 
So should we allow the org admins to do that or should we bring in org concept in actions?

**Review by adithyadinesh0412 — APPROVED** [review]

File: src/controllers/v1/activities.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+	async list(req) {
+		try {
+			let userId = req.decodedToken.id
+			let orgId = req.decodedToken.organization_id
+			if (utils.validateRoleAccess(req.decodedToken.roles, common.ADMIN_ROLE)) {
+				userId = req.query.user_id ? req.query.user_id : req.decodedToken.id
+				orgId = req.query.organization_id ? req.query.organization_id : req.decodedToken.id
+			} else if (utils.validateRoleAccess(req.decodedToken.roles, common.ORG_ADMIN_ROLE)) {
+				userId = req.query.user_id ? req.query.user_id : req.decodedToken.id
```
Comment (aks30) [inline src/controllers/v1/activities.js]: @priyanka-TL why do we need this condition? Any role who has this permission should be able to access this API

File: src/controllers/v1/projects.js [hunk 10]
```diff
@@ -23,7 +22,11 @@ module.exports = class Projects {
 			if (req.params.id) {
 				let project = {}
 				if (req.method === common.REQUEST_METHOD_DELETE) {
-					project = await projectService.delete(req.params.id, req.decodedToken.id)
+					project = await projectService.delete(
+						req.params.id,
+						req.decodedToken.id,
+						req.decodedToken.organization_id
```
Comment (aks30) [inline src/controllers/v1/projects.js]: @priyanka-TL Why is this added for activity tracker PR?

Comment (aks30) [reply src/controllers/v1/projects.js]: Because if this is for activity tracker we should take org id from project/resource

File: src/database/migrations/20240813123014-create-activities-table.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+			updated_at: {
+				allowNull: false,
+				type: Sequelize.DATE,
+			},
+			deleted_at: {
+				type: Sequelize.DATE,
+			},
+		})
+		//add index
+		await queryInterface.addIndex('activities', ['organization_id', 'user_id'], {
```
Comment (aks30) [inline src/database/migrations/20240813123014-create-activities-table.js]: @priyanka-TL in which file do we define creating of distribution columns?

File: src/database/models/activities.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+			},
+			user_id: {
+				allowNull: false,
+				type: DataTypes.STRING,
+			},
+			action_id: {
+				allowNull: false,
+				type: DataTypes.INTEGER,
+			},
+			object_id: {
```
Comment (aks30) [inline src/database/models/activities.js:19]: @priyanka-TL should object id be indexed, because for a given resource people might want all the activities

File: src/validators/v1/actions.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+	update: (req) => {
+		req.checkParams('id')
+			.trim()
+			.optional()
+			.notEmpty()
+			.withMessage('id param is empty')
+			.isNumeric()
+			.withMessage('id param is invalid, must be an integer')
+
+		req.checkBody('code')
```
Comment (aks30) [inline src/validators/v1/actions.js]: @priyanka-TL - I agree with @adithyadinesh0412 only what is required or needs to be updated should be passed in update call, in this code might not be required.

File: src/services/actions.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+	 * Create Action
+	 * @method
+	 * @name create
+	 * @param {String} bodyData - action creation data
+	 * @returns {JSON} - action creation response
+	 */
+	static async create(bodyData) {
+		try {
+			const action = await actionQueries.findOne({ code: bodyData.code })
+			if (action?.id) {
```
Comment (aks30) [inline src/services/actions.js]: @priyanka-TL why do we have to do this check? It will automatically fail because of the DB index right?

File: src/controllers/v1/comments.js [hunk 15]
```diff
@@ -46,7 +46,8 @@ module.exports = class comments {
 				req.params.id ? req.params.id : '',
 				req.query.resource_id,
 				req.body,
-				req.decodedToken.id
+				req.decodedToken.id,
+				req.decodedToken.organization_id
```
Comment (aks30) [inline src/controllers/v1/comments.js]: @priyanka-TL do we not need to pick from resource the org id?

File: src/helpers/eventBroadcaster.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+		queryParams = {},
+	}
+) => {
+	try {
+		const endPoints = events.eventEndpoints[action]
+		await Promise.all(
+			endPoints.map((endPoint) => {
+				if (endPoint.type === common.EVENT_TYPE_API) {
+					if (endPoint.method === 'POST')
+						requester.post(endPoint.baseUrl, endPoint.route, headers, requestBody, queryParams)
```
Comment (aks30) [inline src/helpers/eventBroadcaster.js]: @priyanka-TL are we making API calls to to broad cast events? It should rather be internal event broad casting.

E.g.- https://nodejs.org/api/events.html, https://medium.com/deno-the-complete-reference/10-use-cases-of-event-emitters-in-node-js-d88c0b9bb536

File: src/services/comments.js [hunk 17]
```diff
@@ -22,10 +25,28 @@ module.exports = class CommentsHelper {
...
 	 * @param {Object} bodyData - Request Body
 	 * @param {String} userId - User ID
+	 * @param {String} orgId - Organization ID
 	 * @returns {JSON} - comment id
 	 */
-	static async update(commentId = '', resourceId, bodyData, userId) {
+	static async update(commentId = '', resourceId, bodyData, userId, orgId) {
 		try {
+			//validate resource
+			const resource = await resourceQueries.findOne(
```
Comment (aks30) [inline src/services/comments.js:32]: @priyanka-TL how is this related to activity tracker?

File: src/services/projects.js [hunk 18]
```diff
@@ -253,57 +264,65 @@ module.exports = class ProjectsHelper {
...
-				return responses.failureResponse({
-					message: 'PROJECT_NOT_FOUND',
-					statusCode: httpStatusCode.bad_request,
-					responseCode: 'CLIENT_ERROR',
-				})
+				throw new Error('PROJECT_NOT_FOUND')
 			}
+
+			//add user action
+			eventBroadcaster(common.EVENT_ADD_USER_ACTION, {
```
Comment (aks30) [inline src/services/projects.js]: @priyanka-TL Rather then adding like this in multiple places, can we write a hook in model? E.g. - https://github.com/ELEVATE-Project/mentoring/blob/master/src/database/models/modules.js

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 24 Aug

File: src/validators/v1/actions.js (hunk 13 above)
Comment (priyanka-TL) [reply src/validators/v1/actions.js]: @aks30 update API is for create and update action. I made the changes, if id is there code and description i made as optional when id is there

File: src/controllers/v1/comments.js (hunk 15 above)
Comment (priyanka-TL) [reply src/controllers/v1/comments.js]: it is passed for create activity, but its not required now because the review status will be inprogress when click on start button

File: src/helpers/eventBroadcaster.js (hunk 16 above)
Comment (priyanka-TL) [reply src/helpers/eventBroadcaster.js]: @aks30  this is same like mentoring, currently they are using events for internal api calls, now i added a function call also into that, so both api or function call will work, when call the publish api we can use this event

File: src/services/projects.js (hunk 18 above)
Comment (priyanka-TL) [reply src/services/projects.js]: No we cant add this, hence update API is an auto save call, hence activities will add in each call.

File: src/controllers/v1/activities.js (hunk 9 above)
Comment (priyanka-TL) [reply src/controllers/v1/activities.js]: @aks30 i added because currently i gave this permission to admin and org admin, 
org admin can see any user activity, admin can see any org any user activity.

i will remove it and add permission to all scp users

File: src/services/comments.js (hunk 17 above)
Comment (priyanka-TL) [reply src/services/comments.js:32]: Not a part of activity, It was part of validation PR by mistake committed to here, to validate the status of resource

File: src/services/projects.js (hunk 18 above)
Comment (aks30) [reply src/services/projects.js]: @priyanka-TL we can do for comments model and for resource model we can do the same by checking status, if status not equal to draft or something

File: src/helpers/eventBroadcaster.js (hunk 16 above)
Comment (aks30) [reply src/helpers/eventBroadcaster.js]: @priyanka-TL I don't think the mentoring service uses events, user service might and they do API calls because they have to update the mentoring service if a user or org is updated. We don't need all that here, keep it to what we need only.

File: src/database/models/resources.js [hunk 19]
```diff
@@ -92,6 +94,38 @@ module.exports = (sequelize, DataTypes) => {
...
+					orgId: instance.organization_id,
+				})
+			}
+		} catch (error) {
+			console.error(`Error during ${actionType} hook:`, error)
+			throw error
+		}
+	}
+
+	Resource.addHook('afterCreate', (instance) => emitUserAction(instance, 'RESOURCE_CREATED'))
```
Comment (adithyadinesh0412) [inline src/database/models/resources.js:115]: Can we have RESOURCE_CREATED and RESOURCE_DELETED from common? @priyanka-TL

File: src/services/actions.js [hunk 20]
```diff
@@ -20,11 +20,6 @@ module.exports = class ActionsHelper {
 	 */
 	static async create(bodyData) {
 		try {
-			const action = await actionQueries.findOne({ code: bodyData.code })
```
Comment (adithyadinesh0412) [inline src/services/actions.js]: Can we check if the is already created? @priyanka-TL

File: src/database/models/resources.js (hunk 19 above)
Comment (adithyadinesh0412) [reply src/database/models/resources.js:115]: As discussed resolving this

**Review by adithyadinesh0412 — APPROVED** [review]

File: src/controllers/v1/activities.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+	 * Get user activities
+	 * @method
+	 * @name list
+	 * @param {String} req.pageNo - Page No.
+	 * @param {String} req.pageSize - Page size limit.
+	 * @param {String} req.searchText - Search text.
+	 * @returns {JSON} - activities List.
+	 */
+
+	async list(req) {
```
Comment (aks30) [inline src/controllers/v1/activities.js:22]: @priyanka-TL To include resource id filter

File: src/database/models/actions.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+			tableName: 'actions',
+			freezeTableName: true,
+			paranoid: true,
+		}
+	)
+
+	Action.addHook('beforeDestroy', async (instance, options) => {
+		try {
+			// Soft-delete associated Activities records with matching actions
+			await sequelize.models.Activity.update(
```
Comment (aks30) [inline src/database/models/actions.js]: @priyanka-TL We will not delete any activity if user action is deleted. In action delete API check if there is any activity for that action, if yes then don't allow to delete that action.

File: src/services/reviews.js [hunk 23]
```diff
@@ -260,7 +259,7 @@ module.exports = class reviewsHelper {
 
 			// Publish resource if isPublishResource is true
 			if (isPublishResource) {
-				const publishResource = await this.publishResource(resourceId, resource.user_id)
+				const publishResource = await this.publishResource(resourceId, resource.user_id, orgId)
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL to remove here and everywhere else where org is not required and can be derived from the resource in the hook of the model?

File: src/distributionColumns.psql [hunk 24]
```diff
@@ -9,4 +9,5 @@ SELECT create_distributed_table('resources', 'organization_id');
 SELECT create_distributed_table('resource_creator_mapping', 'creator_id');
 SELECT create_distributed_table('review_resources', 'reviewer_id');
 SELECT create_distributed_table('review_stages', 'organization_id');
-SELECT create_distributed_table('reviews', 'organization_id');
\ No newline at end of file
+SELECT create_distributed_table('reviews', 'organization_id');
+SELECT create_distributed_table('activities', 'organization_id');
```
Comment (aks30) [inline src/distributionColumns.psql]: @priyanka-TL Please add resource id as DC and make corresponding changes

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 29 Aug

File: src/controllers/v1/activities.js (hunk 21 above)
Comment (priyanka-TL) [reply src/controllers/v1/activities.js:22]: added resource id as filter

File: src/database/models/actions.js (hunk 22 above)
Comment (priyanka-TL) [reply src/database/models/actions.js]: done. Removed hook

File: src/services/reviews.js (hunk 23 above)
Comment (priyanka-TL) [reply src/services/reviews.js]: removed

File: src/distributionColumns.psql (hunk 24 above)
Comment (priyanka-TL) [reply src/distributionColumns.psql]: yes done

File: src/validators/v1/activities.js [hunk 25]
```diff
@@ -0,0 +1,18 @@
+/**
+ * name : validators/v1/activities.js
+ * author : Priyanka Pradeep
+ * Date : 29-Aug-2024
+ * Description : Validations of activities controller
+ */
+
+module.exports = {
+	list: (req) => {
```
Comment (aks30) [inline src/validators/v1/activities.js:9]: @priyanka-TL - Activity list will always be with resource id which is mandatory

File: src/database/models/reviews.js [hunk 26]
```diff
@@ -55,5 +57,45 @@ module.exports = (sequelize, DataTypes) => {
 		}
 	)
 
+	Review.addHook('afterSave', async (instance, options) => {
+		try {
+			//get resource type
+			const resource = await sequelize.models.Resource.findOne(
```
Comment (aks30) [inline src/database/models/reviews.js:63]: @priyanka-TL why do we have to make one more query, won't the instance object have org id and type?

File: src/database/models/reviews.js [hunk 27]
```diff
@@ -55,5 +57,45 @@ module.exports = (sequelize, DataTypes) => {
...
+				{
+					where: { id: instance.resource_id },
+					attributes: ['id', 'organization_id', 'type'],
+				},
+				{ raw: true }
+			)
+
+			if (resource?.id) {
```
Comment (aks30) [inline src/database/models/reviews.js]: @priyanka-TL in which case resource id won't be there

File: src/database/migrations/20240813123014-create-activities-table.js [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+			},
+			updated_at: {
+				allowNull: false,
+				type: Sequelize.DATE,
+			},
+			deleted_at: {
+				type: Sequelize.DATE,
+			},
+		})
+		//add index
```
Comment (aks30) [inline src/database/migrations/20240813123014-create-activities-table.js:47]: @priyanka-TL why do we need orgid and user id as index? just object ID is enough that is the only thing will be queries with.

File: src/services/projects.js [hunk 29]
```diff
@@ -729,6 +725,14 @@ module.exports = class ProjectsHelper {
 			}
 
 			await resourceQueries.updateOne({ id: projectData.id }, resourcesUpdate)
+			//add user action
+			eventEmitter.emit(common.EVENT_ADD_USER_ACTION, {
```
Comment (aks30) [inline src/services/projects.js:729]: @priyanka-TL why is this added here and not via hook ?

File: src/services/activities.js [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+			const options = {
+				offset,
+				limit,
+			}
+
+			let filter = {
+				organization_id: orgId.toString(),
+			}
+
+			if (resourceId) {
```
Comment (aks30) [inline src/services/activities.js]: @priyanka-TL if resource id is not given we should not query, without DC key the query would fail right?

File: src/services/actions.js [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+	 */
+
+	static async delete(id) {
+		try {
+			const action = await actionQueries.findById(id)
+			if (!action) {
+				throw new Error('ACTION_NOT_FOUND')
+			}
+
+			const activities = await activitiesQueries.findAllActivities(
```
Comment (aks30) [inline src/services/actions.js:156]: @priyanka-TL won't this query be non-optimal? Without DC and index?

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 29 Aug

File: src/services/actions.js (hunk 31 above)
Comment (priyanka-TL) [reply src/services/actions.js:156]: @aks30 we don't have another option here. When deleting the action, we can't include the org ID. We'll need to verify if any other organisation has activities linked to the action being deleted. Its deletion so its not a frequent activity.

File: src/services/activities.js (hunk 30 above)
Comment (priyanka-TL) [reply src/services/activities.js]: @aks30, I'm not fully convinced about adding object_id as DC. If we also include comments in the future, this might not be the best approach. Additionally, for the list, we won't be able to see all user actions for an organisation, which feels like a limitation.

However, I've made the changes as per your suggestion

File: src/services/projects.js (hunk 29 above)
Comment (priyanka-TL) [reply src/services/projects.js:729]: @aks30, as we've discussed before, once the user submits the resource for a second time or more, we can't identify that through the hook. The first submission sets the resource status to 'submitted,' but for subsequent submissions, it changes to 'in_review'.

File: src/database/migrations/20240813123014-create-activities-table.js (hunk 28 above)
Comment (priyanka-TL) [reply src/database/migrations/20240813123014-create-activities-table.js:47]: @aks30 previously i have designed like org admin can see all users actions, but as per the suggestion i have removed that. 
hence removing the primary key also

File: src/database/models/reviews.js (hunk 27 above)
Comment (priyanka-TL) [reply src/database/models/reviews.js]: @aks30 just to reverify its added, i have removed that

File: src/database/models/reviews.js (hunk 26 above)
Comment (priyanka-TL) [reply src/database/models/reviews.js:63]: @aks30, the reviews table doesn't include the resource type. We'll need to query the resource table to get the resource type.

File: src/validators/v1/activities.js (hunk 25 above)
Comment (priyanka-TL) [reply src/validators/v1/activities.js:9]: done

File: src/services/activities.js [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+	 * @name list
+	 * @param {Integer} resourceId - The ID of the resource for which activities are listed.
+	 * @param {String} userId - Id of the logged in user
+	 * @param {Integer} page - Page number
+	 * @param {Integer} page - Page size limit
+	 * @param {String} search - Search text.
+	 * @returns {JSON} - activities list response.
+	 */
+
+	static async list(resourceId, userId, orgId, page, limit) {
```
Comment (aks30) [inline src/services/activities.js:25]: @priyanka-TL where do we check if object id is not null?

File: src/database/models/reviews.js (hunk 26 above)
Comment (aks30) [reply src/database/models/reviews.js:63]: @priyanka-TL In that case can we try where we are calling review save or update we set resource type as well, it might not save in DB but here in instance object we might get that.

**Review by aks30 — CHANGES_REQUESTED** [review]
Added comments on 30 Aug

File: src/database/models/reviews.js (hunk 26 above)
Comment (priyanka-TL) [reply src/database/models/reviews.js:63]: @aks30 After updating the reviews table we wont get type if we passed also

File: src/services/activities.js (hunk 32 above)
Comment (priyanka-TL) [reply src/services/activities.js:25]: @aks30 In validator we are checking

Comment (priyanka-TL) [reply src/services/activities.js:25]: Added in service level too
